Update PR template for VISC projects

kelliemac opened the pull request:
## Description

Updates PR template to be more comprehensive and better organized, so that people are reminded "just-in-time" of the key steps they should be taking before each submitting a PR for review (within an analysis repo).

In particular:

- I exposed the guidance text that was previously in comments so that people would be more likely to pay attention to and modify it.
- I copied the checklists from VISC-documentation directly into the PR template, and added/reorganized/simplified to make the checklists more comprehensive but hopefully not too overwhelming.

## Related Issues

Related to https://github.com/FredHutch/VISCtemplates/issues/228 and in some sense also https://github.com/FredHutch/VISCtemplates/issues/119 and https://github.com/FredHutch/VISCtemplates/issues/181

## Checklist

- [ ] This PR includes unit tests
- [ ] This PR establishes a new function or updates parameters in an existing function
  - [ ]  The roxygen skeleton for this function has been updated using `devtools::document`
- [x] I have updated NEWS.md to describe the proposed changes

kelliemac committed:
Copy checklists directly into pull_request_template.md

kelliemac committed:
change master to main

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
switch order of code review and writing review

kelliemac committed:
add documentation section

kelliemac committed:
add note about when/who to complete checklist

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
add link to VISCtemplates and VISCfunctions repos

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update NEWS.md

kelliemac commented:
@asatofh and @mayerbry I took a stab at updating the analysis repo PR template to be more in line with the actual workflow I use and tasks I know need to be addressed. hoping this will help the PR process become a little smoother for both PR creators and reviewers. any feedback is welcome!!!

mayerbry commented on inst/templates/pull_request_template.md:
```diff
@@ -1,21 +1,94 @@
+See also [code review guidelines](https://github.com/FredHutch/VISC-Documentation/blob/main/Programming/code-review-guideline.md)
+
+- [ ]  I have compiled the R Markdown file(s) (or run the relevant code) with no errors
+    - [ ]  Running time has been recorded or estimated: ___________
```
```suggestion
    - [ ]  Warnings are not suppressed. If a warnings must be suppressed there is a clear explanation (i.e., comment).
    - [ ]  Running time has been recorded or estimated: ___________
```

mayerbry commented on inst/templates/pull_request_template.md:
```diff
@@ -1,21 +1,94 @@
+- [ ] Necessary context for the project/analysis has been documented
+- [ ] Appropriate README.md files have been updated to reflect the latest changes
+- [ ] The latest versions of all relevant files have been pushed to the repo
+    - [ ] Unrelated or unnescessary files aren't included (e.g., LaTeX .toc files)
```
```suggestion
    - [ ] Unrelated or unnecessary files aren't included (e.g., LaTeX .toc files)
```

mayerbry commented on inst/templates/pull_request_template.md:
```diff
@@ -1,21 +1,94 @@
+    - [ ] File paths are relative (except for trials and network drive paths) and portable across operating systems (use `file.path()`)
+    - [ ] Functions are organized and well-documented, with explanations of purpose, inputs, and ouput
+    - [ ] Hard coding and magic numbers are avoided
+    - [ ] Comments are helpful and do not include debt (e.g. `# TODO:`) or commented-out backup code
```
```suggestion
    - [ ] Comments are helpful and do not include unaddressed debt (e.g. `# TODO:`)
    - [ ] Commented-out backup code and unused chunks have been removed
```

mayerbry commented on inst/templates/pull_request_template.md:
```diff
@@ -1,21 +1,94 @@
+    - [ ] I've opened and reviewed the compiled PDF document
+    - [ ] I've opened and reviewed the compiled Word document
+- [ ] There are no obvious Markdown/pandoc/Latex errors 
+    - [ ] No broken references (?? or ???) in the text (Use find: “??”)
```
```suggestion
    - [ ] No broken references (?? or ???) in the text (Use find: “??”)
    - [ ] No stray warnings or R output in the text (Use find: “#”)
```

mayerbry commented on inst/templates/pull_request_template.md:
```diff
@@ -1,21 +1,94 @@
+- [ ] I have reviewed the results carefully
+    - [ ] Everything mentioned in the Summary of Main Results is also in the Results section
+    - [ ] Statements in Results section are correct (including p-values) and supported by the correct figure and table references
+        - [ ] Variables are used in referring to numeric values in the Results section (to minimize human error)
```
I'd prefer "inline referencing" or something similar sentiment (code-based?). I often use filtering commands and do not create variables

mayerbry reviewed (approved):
Thanks Kellie! some super minor comments from me

kelliemac committed:
fix type

Co-authored-by: Bryan Mayer <39173901+mayerbry@users.noreply.github.com>

kelliemac committed:
split two checklist items apart

Co-authored-by: Bryan Mayer <39173901+mayerbry@users.noreply.github.com>

kelliemac committed:
add to checklist searching for stray warnings and R output

Co-authored-by: Bryan Mayer <39173901+mayerbry@users.noreply.github.com>

kelliemac committed:
add suppressed warnings to checklist

Co-authored-by: Bryan Mayer <39173901+mayerbry@users.noreply.github.com>

kelliemac committed:
change wording about inserting numeric values in results section

kelliemac commented:
thanks @mayerbry! I incorporated all your suggestions. @asatofh do you want to take a look too, or are you happy for me to merge this PR?

kelliemac committed:
Merge branch 'develop' into update-pr-templates

asatofh commented:
> thanks @mayerbry! I incorporated all your suggestions. @asatofh do you want to take a look too, or are you happy for me to merge this PR?

The main thing I would love to see is more / better instructions to the reviewer about what specifically to review. Something like (highlighted in bold italics for your review):

Here, describe your changes in detail **_and provide instruction to the reviewer on areas on which you'd like them to focus_**. Give a short background on the report, outline important questions or details for the reviewer, and add links to any supporting documents (e.g., protocol presentation). It's helpful to add links to the key files with the unique ID for the commit (a.k.a. the "SHA" or "hash").

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac commented:
> > thanks @mayerbry! I incorporated all your suggestions. @asatofh do you want to take a look too, or are you happy for me to merge this PR?
> 
> @kelliemac The main thing I would love to see is more / better instructions to the reviewer about what specifically to review. I know it's already stated, but often forgotten! Something like (highlighted in bold italics for your review):
> 
> Here, describe your changes in detail **_and provide instruction to the reviewer on areas on which you'd like them to focus_**. Give a short background on the report, outline important questions or details for the reviewer, and add links to any supporting documents (e.g., protocol presentation). It's helpful to add links to the key files with the unique ID for the commit (a.k.a. the "SHA" or "hash").

@asatofh I added a section specifically with a checklist for the reviewer. let me know what you think!

kelliemac committed:
add note about data package version

slager committed:
Merge branch 'develop' into update-pr-templates

mayerbry commented:
## Analysis Code Checklist

- [ ] Knit .Rmd yourself to see if everything runs correctly.
- [ ] Check joins and filtering of data.
- [ ] Code is readable, easy to understand, and well-documented.
- [ ] Code follows the style guide.
- [ ] Code follows the functional code guide.
- [ ] Code follows what is written in the statistical methods section.
- [ ] You could own (e.g., explain and reproduce) this code if you had to.


## Writing Review Checklist

- [ ] Does everything look right in the reproducibility tables?
- [ ] Are the tables sorted to present the results in parallel with the Results Section?
- [ ] Double-check significance highlighting.
- [ ] Do footnotes, captions, x- and y- axes, and data in the plot all match / make sense?
- [ ] Plots: do you have at least 2 y-axis tick marks? Do your tick marks cover the minimum and maximum values displayed?
- [ ] Do your captions match the figure and table results?
- [ ] Is the sample type accurate (e.g., serum, plasma, PBMC)? If unsure, confirm with Michelle/Eva.
- [ ] Any broken references (?? or ???) in text (Use find: “??”)?
- [ ] Are the references to figures and tables correct? Do they support what is written in the text?
- [ ] Did you spell-check the text (Hotkey: F7 or `spelling` package)?
- [ ] Did you check that the spelling in all the captions and footnotes (RStudio spellcheck will not check these)?
- [ ] Past tense consistently throughout (Use Find: “are”)
- [ ] Do significant p-values match what results you’re seeing in the plot?
- [ ] Once completed have you committed and pushed all relevant files for the report (i.e. bib file, …)? Can someone else compile this if they cloned the repo?
- [ ] If any iterative changes were made to plots/figures/included data/grammar, search relevant terms in the text to make sure changes were reflected throughout the report.
- [ ] Post revisions: have you updated your captions?
- [ ] Use git diff on the .tex file to confirm differences between original and new revision.

slager committed:
Merge branch 'develop' into update-pr-templates

slager commented:
This is just a demonstration comment that I will delete soon.

The code quoted below is unrelated to the changes in this PR.

https://github.com/FredHutch/VISCtemplates/blob/91dc8a7b405852b4735aa6da0920001f6f527dff/R/create_visc_project.R#L112-L119

kelliemac committed:
incorporate suggestions from Claudio

kelliemac committed:
Update description

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
Update pull_request_template.md

kelliemac committed:
update report pdf review

kelliemac committed:
Merge branch 'develop' into update-pr-templates

kelliemac committed:
Update NEWS.md

kelliemac commented:
@asatofh @mayerbry and @lemireg I made some updates based on the document that Claudio put together. can you review and let me know what you think?

Commit:
merge develop into branch

Commit:
standardize spacing in NEWS.ms

kelliemac committed:
further updates to pr template

kelliemac committed:
finish drafting full shortened checklist

kelliemac committed:
Merge branch 'develop' into update-pr-templates

kelliemac commented:
@asatofh, @mayerbry, @lemireg, @asatofh will you take another look at this updated PR template when you have a chance?

https://github.com/FredHutch/VISCtemplates/blob/update-pr-templates/inst/templates/pull_request_template.md

I simplified the template recently relative to my original changes - basically, I'm trying to strike the right balance between pointing out key things that need to be checked and not creating an overwhelmingly long list that people will just ignore.

lemireg commented:
Thank you so much for doing this @kelliemac ! At some point we could flag this for Claudio to look at and/or review as well since it's something he's been thinking about and may have feedback. A few comments or questions:

- I like your implementation suggestions in brackets (ex: (Use find: "??") or (Use find: "#")... next to "Text has been spell-checked, including captions and footnotes" we could add (F7 in R studio runs spell-check). @mayerbry is this the spell-check you run? or is it an outside package loaded in?)
- What do you mean by this: "No local installations"? Of R-packages? Local as in non-CRAN? Or no datapackages located on local (would you be able to see that in the code?)?
- line 66 unnescessary --> unnecessary
- Can we clarify what this means: " Statistical analysis code matches what is described and intended" maybe we can add " Statistical analysis code matches what is described and intended in the Statistical Methods section of the report" I am suggesting this to clarify the difference between this point and this other point lower down: "Agreement between the report and the corresponding SAP and study protocol"
- Following up on that last point about checking the SAP, that feels like a separate category from Basic, Code and Report. The three lists are all about internal consistency and coding/report standards. Those reviews could be done without any other study or project level context. However, matching the analysis to the SAP is a heavier lift and requires access to the SAP (if we want this included, then PR Creator should maybe be asked to link to the SAP location in their PR?). I would be interested in others' thoughts on this!

kelliemac commented:
Turning suggestions from Gabby into a list of to-do items so I can track them:

- [x] Add "F7 in R studio runs spell-check" or similar
- [x] Clarify what is meant by "No local installations" and how this relates to data packages
- [x] line 66 unnescessary --> unnecessary
- [x] Clarify "Statistical analysis code matches what is described and intended"
- [x] Consider removing or modifying point about checking the SAP and protocol

kelliemac committed:
Fix typo "unnecessary"

kelliemac commented:
related to spell check: I typically do this in Word, because I think this captures more details than checking in Rstudio (e.g., captions and values within tables). curious to hear what others prefer and recommend, though.

kelliemac committed:
update mentions of SAP

Updated the pull request template to improve clarity and provide specific guidance for reviewers.

kelliemac committed:
add details for spell check and local installations

kelliemac committed:
clarify code logic wording and add item to check if report has been compiled

kelliemac committed:
Merge branch 'develop' into update-pr-templates

kelliemac commented:
@lemireg @mayerbry @asatofh @slager I updated based on our conversation today! please take one more look at https://github.com/FredHutch/VISCtemplates/blob/update-pr-templates/inst/templates/pull_request_template.md and either approve PR or request additional changes!

mayerbry reviewed (approved):
thanks for doing this!
